lxd/storage/drivers/driver_alletra_wsapi1.go

```diff
@@ -11,3 +40,237 @@ func newAlletraClient(driver *alletra) *alletraClient {
+			},
+		},
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("Failed to send request: %w", err)
+	}
+
+	defer resp.Body.Close()
+
+	var responseBodyBuffer bytes.Buffer
+	teeReader := io.TeeReader(resp.Body, &responseBodyBuffer)
+	bodyBytes, err := io.ReadAll(teeReader)
+	if err != nil {
+		return fmt.Errorf("Failed to read response body for TeeReader: %w", err)
+	}
+
+	if resp.Header.Get("Content-Type") != "application/json" && len(bodyBytes) > 0 {
+		return fmt.Errorf("Response Content-type: %q. Only application/json is allowed for non-empty response body", resp.Header.Get("Content-Type"))
+	}
+
+	if resp.StatusCode >= 400 {
+		respBody = &hpeError{}
+		err := json.Unmarshal(bodyBytes, respBody)
+		if err != nil {
+			return fmt.Errorf("HPE failed to parse WSAPI error response: %w", err)
+		}
+	}
+
+	// Extract the response body if requested.
+	if len(bodyBytes) > 0 {
+		err = json.Unmarshal(bodyBytes, &respBody)
+		if err != nil {
+			return fmt.Errorf("HPE failed to read response body from %q: %w", url.String(), err)
+		}
+	}
+
+	// Extract the response headers if requested.
+	if respHeaders != nil {
+		for k, v := range resp.Header {
+			respHeaders[k] = strings.Join(v, ",")
+		}
+	}
+
+	// Return the formatted error from the body
+	hpeErr, assert := respBody.(*hpeError)
+	if assert {
+		hpeErr.HTTPStatusCode = resp.StatusCode
+
+		// The unauthorized error is reported when an invalid (or expired) access token is provided.
+		// Wrap unauthorized requests into an API status error to allow easier checking for expired
+		// token in the requestAuthenticated function.
+		if resp.StatusCode == http.StatusForbidden && hpeErr.Code == 6 { // 6 = INV_SESS_KEY
+			return api.StatusErrorf(http.StatusForbidden, "Unauthorized request")
+		}
+
+		return hpeErr
+	}
+
+	return nil
+}
+
+// requestAuthenticated issues an authenticated HTTP request against the HPE Storage gateway.
+// In case the Session Key is expired, the function will try to obtain a new one.
+func (p *alletraClient) requestAuthenticated(method string, url url.URL, reqBody map[string]any, respBody any) error {
+	// If request fails with an unauthorized error, the request will be retried after
+	// requesting a new session key.
+	retries := 1
+
+	for {
+		// Ensure we are logged into the WSAPI.
+		err := p.login()
+		if err != nil {
+			return err
+		}
+
+		// Add Session Key to Headers.
+		reqHeaders := map[string]string{
+			"X-HP3PAR-WSAPI-SessionKey": p.sessionKey,
+		}
+
+		// Initiate request.
+		err = p.request(method, url, reqBody, reqHeaders, &respBody, nil)
+		if err != nil {
+			if api.StatusErrorCheck(err, http.StatusForbidden) && retries > 0 {
+				// Session key seems to be expired.
+				// Reset the token and try one more time.
+				p.sessionKey = ""
+				retries--
+				continue
+			}
+
+			hpeErr, assert := err.(*hpeError)
+			if assert {
+				logger.Debugf("HPE debug hpeErr.statusCode: %d", hpeErr.HTTPStatusCode)
```

Comment:
Lets remove these debug statements, or if they are needed, lets switch to a single contextual log line.

Reply:
Have done!